core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/kopt/KOptDescriptor.java

```diff
@@ -0,0 +1,504 @@
+     *        argument is between its first and last argument when the tour is
+     *        taken in the successor direction.
+     */
+    public KOptDescriptor(
+            Node_[] removedEdges,
+            int[] addedEdgeToOtherEndpoint,
+            Function<Node_, Node_> endpointToSuccessorFunction,
+            TriPredicate<Node_, Node_, Node_> betweenPredicate) {
+        this.k = (removedEdges.length - 1) >> 1;
+        this.removedEdges = removedEdges;
+        this.removedEdgeIndexToTourOrder = new int[removedEdges.length];
+        this.inverseRemovedEdgeIndexToTourOrder = new int[removedEdges.length];
+        this.addedEdgeToOtherEndpoint = addedEdgeToOtherEndpoint;
+
+        // Compute the permutation as described in FindPermutation
+        // (Section 5.3 "Determination of the feasibility of a move",
+        //  An Effective Implementation of K-opt Moves for the Lin-Kernighan TSP Heuristic)
+        for (int i = 1, j = 1; j <= k; i += 2, j++) {
+            removedEdgeIndexToTourOrder[j] =
+                    (endpointToSuccessorFunction.apply(removedEdges[i]) == removedEdges[i + 1]) ? i : i + 1;
+        }
+
+        IntComparator comparator = (pa, pb) -> pa == pb ? 0
+                : (betweenPredicate.test(removedEdges[removedEdgeIndexToTourOrder[1]], removedEdges[pa], removedEdges[pb]) ? -1
+                        : 1);
+        IntArrays.stableSort(removedEdgeIndexToTourOrder, 2, k + 1, comparator);
+
+        for (int i, j = 2 * k; j >= 2; j -= 2) {
+            removedEdgeIndexToTourOrder[j - 1] = i = removedEdgeIndexToTourOrder[j / 2];
+            removedEdgeIndexToTourOrder[j] = ((i & 1) == 1) ? i + 1 : i - 1;
+        }
+
+        for (int i = 1; i <= 2 * k; i++) {
+            inverseRemovedEdgeIndexToTourOrder[removedEdgeIndexToTourOrder[i]] = i;
+        }
+    }
+
+    public int[] getRemovedEdgeIndexToTourOrder() {
+        return removedEdgeIndexToTourOrder;
+    }
+
+    /**
+     * Calculate the disjoint k-cycles for {@link KOptDescriptor#removedEdgeIndexToTourOrder}. <br />
+     * <br />
+     * Any permutation can be expressed as combination of k-cycles. A k-cycle is a sequence of
+     * unique elements (p_1, p_2, ..., p_k) where
+     * <ul>
+     * <li>p_1 maps to p_2 in the permutation</li>
+     * <li>p_2 maps to p_3 in the permutation</li>
+     * <li>p_(k-1) maps to p_k in the permutation</li>
+     * <li>p_k maps to p_1 in the permutation</li>
+     * <li>In general: p_i maps to p_(i+1) in the permutation</li>
+     * </ul>
+     * For instance, the permutation
+     * <ul>
+     * <li>1 -> 2</li>
+     * <li>2 -> 3</li>
+     * <li>3 -> 1</li>
+     * <li>4 -> 5</li>
+     * <li>5 -> 4</li>
+     * </ul>
+     * can be expressed as `(1, 2, 3)(4, 5)`.
+     *
+     * @return The {@link KOptCycleInfo} corresponding to the permutation described by
+     *         {@link KOptDescriptor#removedEdgeIndexToTourOrder}.
+     */
+    public KOptCycleInfo getCyclesForPermutation() {
```

Comment:
Let's hide all the methods that aren't strictly necessary for the class's API.

(You may have already noticed I am fairly strict about this sort of thing - there is a reason. Everything you expose, people will use and abuse. Everything you expose, you will pay for eventually. My experience teaches me you don't want to expose things unless you have to.)

Reply:
Also, it appears to me as if the data this method operates on never really change.
Why not cache the result then? 

Instances of this class are fairly high in their count on the heap, yet it seems to me as if there should really only be one. Am I missing something?